File: Graph/detect_cycle_und_grp_bfs.cpp
class Solution {
    private:
    bool isCycleBFS(vector<vector<int>> &adj, vector<bool> &vis, int V, int n,int i, int parent){
        vis[i]=true;
        queue<pair<int,int>> q;
        q.push({i,parent});
        while(!q.empty()){
            int child = q.front().first, prnt=q.front().second;
            q.pop();
            for(auto &it: adj[child]){
                if(it==prnt)
                continue;
                else if(vis[it])
                return true;
                else
                {
                    q.push({it,child});
                    vis[it]=true;
                }
            }
        }
        return false;
    }
  public:
    bool isCycle(int V, vector<vector<int>>& edges) {
        // Code here
        vector<vector<int>> adj(V);
        int n = edges.size();
        for(auto &edge:edges){ // can also write (auto edge : edges) but it will create copy, so extra memory usage
            int u = edge[0];
            int v = edge[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        vector<bool> vis(V,false);
         for(int i=0; i<V; i++){
             if(!vis[i] && isCycleBFS(adj,vis,V,n,i,-1))
             return true;
         }
         return false;
    }
};

File: Graph/detect_cycle_und_grp_bfs.java
import java.util.*;

class Solution {
    boolean isCycleBfs(List<List<Integer>> adj, boolean[] vis, int node, int parent){
        vis[node]=true;
        Queue<int []> q = new LinkedList<>();
        q.add(new int[]{node,parent});
        while(!q.isEmpty()){
            int[] pair = q.poll();
            int child = pair[0];
            int pr = pair[1];
            for(int i: adj.get(child)){
                if(i==pr)
                continue;
               else if(vis[i])
                return true;
                else{
                    vis[i] = true;
                    q.add(new int[]{i,child});
                }
            }
        }
        return false;
    }
    public boolean isCycle(int V, int[][] edges) {
        boolean[] vis = new boolean[V];
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0; i<V; i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0; i<edges.length; i++){
            int u= edges[i][0];
            int v = edges[i][1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        for(int i=0; i<V; i++){
            if(!vis[i] && isCycleBfs(adj,vis,i,-1))
            return true;
        }
        return false;
    }
}